### PR TITLE
Add public API `WriteWithCallback` to support custom callbacks

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR adds a `DB::WriteWithCallback` API that does the same things as `DB::Write` while takes an argument `UserWriteCallback` to execute custom callback functions during the write.

We currently support two types of callback functions: `OnWriteEnqueued` and `OnWalWriteFinish`. The former is invoked   after the write is enqueued, and the later is invoked after WAL write finishes when applicable.

These callback functions are intended for users to use to improve synchronization between concurrent writes, their execution is on the write's critical path so it will impact the write's latency if not used properly. The documentation for the callback interface mentioned this and suggest user to keep these callback functions' implementation minimum.

Although transaction interfaces' writes doesn't yet allow user to specify such a user write callback argument, the `DBImpl::Write*` type of APIs do not differentiate between regular DB writes or writes coming from the transaction layer when it comes to supporting this `UserWriteCallback`. These callbacks works for all the write modes including: default write mode, Options.two_write_queues, Options.unordered_write, Options.enable_pipelined_write

Test plan:
Added unit test in ./write_callback_test

### CODE REVIEW

**Review by anand1976 — COMMENTED** [review]
How does the user specify the callback? In `WriteOptions` perhaps?

**Comment by jowlyzhang** [comment]
> How does the user specify the callback? In `WriteOptions` perhaps?

Yeah, that's a good question. `WriteOptions` would work, another idea is maybe we can make `WriteWithCallback` a public API too?

**Review by anand1976 — COMMENTED** [review]
LGTM! Thanks for the PR and quick turnaround

**Comment by jowlyzhang** [comment]
> LGTM! Thanks for the PR and quick turnaround

Thank you for the quick review!

File: include/rocksdb/user_write_callback.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "rocksdb/status.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Custom callback functions to support users to plug in logic while data is
+// being written to the DB. It's intended for better synchronization between
+// concurrent writes. Note that these callbacks are in the write's critical path
+// It's desirable to keep them fast and minimum to not affect the write's
+// latency.
```
Comment (anand1976) [inline include/rocksdb/user_write_callback.h:17]: Also mention they (atleast the post WAL callback) may be called in the context of a different thread.

File: utilities/transactions/write_prepared_txn.cc [hunk 2]
```diff
@@ -248,7 +249,7 @@ Status WritePreparedTxn::CommitInternal() {
   // true. See the comments about GetCommitTimeWriteBatch() in
   // include/rocksdb/utilities/transaction.h.
   s = db_impl_->WriteImpl(write_options_, working_batch, nullptr, nullptr,
-                          zero_log_number, disable_memtable, &seq_used,
+                          nullptr, zero_log_number, disable_memtable, &seq_used,
```
Comment (anand1976) [inline utilities/transactions/write_prepared_txn.cc:252]: For the sake of consistency, can we say `/*user_write_cb=*/ nullptr`?

**Review by anand1976 — APPROVED** [review]
LGTM. A couple of minor nits.

File: include/rocksdb/user_write_callback.h (hunk 1 above)
Comment (jowlyzhang) [reply include/rocksdb/user_write_callback.h]: Thank you for the suggestion!

**Pull request closed without merging**